File: #10472 십자뒤집기/구본관.py
import sys
from collections import deque

input = sys.stdin.readline

dydx = [[0,0],[-1,0],[1,0],[0,-1],[0, 1]]

point = []
for y in range(3):
    for x in range(3):
        point.append([y, x])

dic = {}

dq = deque()
dq2 = deque()
dq2.append(0)
num = 0

def paint(bit, y, x):
    for dy, dx in dydx:
        ny = y + dy
        nx = x + dx
        if 0<=nx<3 and 0<=ny<3:
            bit = bit ^ 1 << (8 - point.index([ny,nx]))
    return bit

nbit = paint(0, 1, 0)

while dq2:
    dq = dq2
    dq2 = deque()
    while dq:
        bit = dq.popleft()

        if bit in dic:
            continue
        dic[bit] = num

        for i in range(9):
            y,x = point[i]
            nbit = paint(bit, y,x)

            # b = bin(nbit)[2:]
            # b = b.zfill(9)
            # print(b[:3])
            # print(b[3:6])
            # print(b[6:])
            # print()

            dq2.append(nbit)
    num+=1

for _ in range(int(input())):
    li = "0b"
    for _ in range(3):
        for i in input().strip():
            if i=="*":
                li += "1"
            else:
                li += "0"
    print(dic[int(li,2)])

File: #10472 십자뒤집기/구본관.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	static int[][] dydx;
	static List<int[]> point;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        
        dydx = new int[][] {{0,0},{-1,0},{1,0},{0,-1},{0,1}};
        
        point = new ArrayList<>();
        int index=0;
        for(int y=0; y<3; y++) {
        	for(int x=0; x<3; x++) {
        		point.add(new int[] {y,x});
        	}
        }
        
        HashMap<Integer, Integer> dic = new HashMap<>();
        
        Deque<Integer> dq = new ArrayDeque<>();
        Deque<Integer> dq2 = new ArrayDeque<>();
        dq2.offer(0);
        int num = 0;
        
        int nbit = paint(0,1,0);
        
        while(!dq2.isEmpty()) {
        	dq = dq2;
        	dq2 = new ArrayDeque<>();
        	while(!dq.isEmpty()) {
        		int bit = dq.pollFirst();
        		
        		if(dic.containsKey(bit)) {
        			continue;
        		}
        		dic.put(bit, num);
        		
        		for(int i=0; i<9; i++) {
        			int y = point.get(i)[0];
        			int x = point.get(i)[1];
        			
        			nbit = paint(bit,y,x);
        			
        			dq2.add(nbit);
        		}
        	}
        	num+=1;
        }
        
        int n = Integer.parseInt(br.readLine());
        for(int i=0; i<n; i++) {
        	char[] str = new char[9];
        	index = 0;
        	for(int t=0; t<3; t++) {
        		String s = br.readLine();
        		for(int x=0; x<3; x++) {
        			if(s.charAt(x)=='*') {
        				str[index++] = '1';
        			}else {
        				str[index++] = '0';
        			}
        		}
        	}
        	sb.append(dic.get(Integer.parseInt(new String(str), 2)) + "\n");
        }

        System.out.println(sb);
    }
    static int paint(int bit, int y, int x) {
    	for(int[] yx : dydx) {
    		int ny = y + yx[0];
    		int nx = x + yx[1];
    		
    		if(0<=nx && nx<3 && 0<=ny && ny<3) {
    			for(int i=0; i<9; i++) {
    				if(Arrays.equals(point.get(i), new int[] {ny, nx})) {
    					bit = bit ^ 1<<(8-i);
    					break;
    				}
    			}
    		}
    	}
    	return bit;
    }
}
